lib/charms/kubernetes_charm_libraries/v0/multus.py

```diff
@@ -0,0 +1,399 @@
+        """
+        if not network_annotations:
+            logger.info("No network annotations were provided")
+            return
+        try:
+            statefulset = self.client.get(res=StatefulSet, name=name, namespace=self.namespace)
+        except ApiError:
+            raise KubernetesMultusError(f"Could not get statefulset {name}")
+        statefulset.spec.template.metadata.annotations["k8s.v1.cni.cncf.io/networks"] = json.dumps(  # type: ignore[attr-defined]  # noqa: E501
+            [network_annotation.dict() for network_annotation in network_annotations]
+        )
+        for container_name in containers_requiring_net_admin_capability:
+            container_index = self.get_container_index_from_name(
+                statefulset_name=name, container_name=container_name
+            )
+            statefulset.spec.template.spec.containers[  # type: ignore[attr-defined]
+                container_index
+            ].securityContext.capabilities = Capabilities(
+                add=[
+                    "NET_ADMIN",
+                ]
+            )
+        try:
+            self.client.patch(
+                res=StatefulSet,
+                name=name,
+                obj=statefulset,
+                patch_type=PatchType.MERGE,
+                namespace=self.namespace,
+            )
+        except ApiError:
+            raise KubernetesMultusError(f"Could not patch statefulset {name}")
+        logger.info(f"Multus annotation added to {name} Statefulset")
+
+    def statefulset_is_patched(
+        self,
+        name: str,
+        network_annotations: list[NetworkAnnotation],
+        containers_requiring_net_admin_capability: list[str],
+    ) -> bool:
+        """Returns whether the statefulset has the expected multus annotation.
+
+        Args:
+            name: Statefulset name.
+            network_annotations: list of network annotations
+            containers_requiring_net_admin_capability: Containers requiring NET_ADMIN capability
+
+        Returns:
+            bool: Whether the statefulset has the expected multus annotation.
+        """
+        try:
+            statefulset = self.client.get(res=StatefulSet, name=name, namespace=self.namespace)
+        except ApiError:
+            raise KubernetesMultusError(f"Could not get statefulset {name}")
+        if "k8s.v1.cni.cncf.io/networks" not in statefulset.spec.template.metadata.annotations:  # type: ignore[attr-defined]  # noqa: E501
+            logger.info("Multus annotation not yet added to statefulset")
+            return False
+        if json.loads(
+            statefulset.spec.template.metadata.annotations["k8s.v1.cni.cncf.io/networks"]  # type: ignore[attr-defined]  # noqa: E501
+        ) != [network_annotation.dict() for network_annotation in network_annotations]:
+            logger.info("Existing annotation are not identical to the expected ones")
+            return False
```

Comment:
I think this will crash when the annotation exists but is not valid JSON. Maybe it would be better to catch the exception and return `False` in that case?

Reply:
Done